java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -998,42 +1099,54 @@ class DatabaseImpl(
-                    DataType.Collection.ordinal.toString()
+            if (entitiesAreTopLevel) {
+                // Find all collections with missing entries (collections that were updated).
+                val updatedContainersStorageKeys = rawQuery(
+                    """
+                        SELECT storage_keys.storage_key
+                        FROM storage_keys
+                        LEFT JOIN collection_entries
+                            ON storage_keys.value_id = collection_entries.collection_id
+                        WHERE storage_keys.data_type IN (?, ?)
+                        AND collection_entries.value_id NOT IN (SELECT id FROM entity_refs)
+                    """.trimIndent(),
+                    arrayOf(
+                        DataType.Singleton.ordinal.toString(),
+                        DataType.Collection.ordinal.toString()
+                    )
+                ).map { it.getString(0) }.toSet()
+
+                // Remove from collection_entries all references to the expired entities.
+                delete(
+                    TABLE_COLLECTION_ENTRIES,
+                    """
+                        collection_id IN (SELECT id FROM collections WHERE type_id > ?)
+                        AND value_id NOT IN (SELECT id FROM entity_refs)
+                    """.trimIndent(),
+                    arrayOf(LARGEST_PRIMITIVE_TYPE_ID.toString()) // only entity collections.
                 )
-            ).map { it.getString(0) }.toSet()
 
-            // Remove from collection_entries all references to the expired entities.
-            delete(
-                TABLE_COLLECTION_ENTRIES,
-                """
-                    collection_id IN (SELECT id FROM collections WHERE type_id > ?)
-                    AND value_id NOT IN (SELECT id FROM entity_refs)
-                """.trimIndent(),
-                arrayOf(LARGEST_PRIMITIVE_TYPE_ID.toString()) // only entity collections.
-            )
-            (storageKeys union updatedContainersStorageKeys).map { storageKey ->
-                notifyClients(StorageKeyParser.parse(storageKey)) {
-                    it.onDatabaseDelete(null)
+                println("gotta notify")
+                println(storageKeys.joinToString())
+                println(updatedContainersStorageKeys.joinToString())
+                (storageKeys union updatedContainersStorageKeys).map { storageKey ->
+                    notifyClients(StorageKeyParser.parse(storageKey)) {
+                        it.onDatabaseDelete(null)
+                    }
                 }
+            } else {
+                emptyList()
             }
         }
     }
 
     private fun deleteFields(storageKeyIds: Array<String>, db: SQLiteDatabase) = db.transaction {
         // List of question marks of the same length, to be used in queries.
         val questionMarks = questionMarks(storageKeyIds)
+
+        val collectionFields =
```

Comment:
this and similar vals are used in multiple places, should we extract them to class level vals? it could help efficiency